active/0019-plugins.md

```diff
@@ -0,0 +1,126 @@
+- Callbacks under the same hook point are ordered by their priority, however most of the plugins (internal plugins included) today are using the default value 0, meaning it's ordered by luck.
+
+Worth mentioning, since 4.3.2, EMQ X for the first supported drop-in installation of external plugins. That is, plugins can be compiled and packaged separately (instead of being released as a part of EMQ X package), see [Develop and deploy EMQ X plugin for Enterprise 4.3](https://emqx.atlassian.net/wiki/spaces/EMQX/blog/2021/05/23/168591472)
+
+## References
+
+- RabbitMQ: [Plugin Development Basics — RabbitMQ](https://www.rabbitmq.com/plugin-development.html)
+- Grafana: [Pie Chart](https://grafana.com/grafana/plugins/grafana-piechart-panel/)
+- WordPress: [WordPress Plugins](https://wordpress.org/plugins/)
+- RabbitMQ Plugin release packages: [Releases · rabbitmq/rabbitmq-management-exchange](https://github.com/rabbitmq/rabbitmq-management-exchange/releases)
+
+### High level requirements
+
+1. Hook points should be compatible with 4.x
+2. More clear ordering of all hook callbacks under one certain hook point, for example, internal hooks use priority 1000, and force external plugins to provide a priority number. If they wish to have it ordered before internal plugins they can choose to use a number smaller than 1000.
+3. All built-in features such as authentication and authorisation are not presented as plugins except for
+    1. LDAP auth
+
+### Basic steps to install an external plugin
+
+- Download compiled zip package
+- Upload to a specific dir
+- Execute a command to validate & install & enable the plugin
+
+### Manage plugins from Dashboard UI
+
+- Manage installation from Dashboard GUI
+    - upload
+    - install
+    - uninstall
+- Manage a list of installed plugins, supported actions:
+    - List view
+    - Show running status: running | stopped
+    - support actions: start | stop
+
+## Plugin metadata
+
+The plugin package should include metadata (in JSON format) to help identify, validate, or describe the package.
+
+- Name (same as the Erlang application name, it has to be globally unique)
+- Version
+- License (checked against a white list: apache2, mit, ...)
+- Author
+    - Name
+    - Contact
+- URL to source code
+- What functionalities (one or more of below)
+    - authentication
+    - authorisation
+    - data_persistence
+    - rule_engine_extension
+- Compatibility
+    - Compiled with EMQ X version (from 5.0)
+    - Supported OTP releases (has to be the same as EMQ's supported OTP versions)
+
+## User Interface
+
+### Management **APIs**
+
+- List plugins
+
+```json
+/plugins
+[{"metadata":
+    { "name": "emqx_foobar",
+      "description": "EMQ X plugin to implement foobar feature",
+      "version": "0.1.0",
+      ...
+    },
+  "status": "running" // not_initialized | running | stopped
+},..]
+```
+
+The `not_initialized` state is recognised when the plugin is installed (unziped), but not configured to be loaded.
```

Comment:
Q: is `not_initalized` different from `stopped`?

Reply:
not_initialized is when the package is extracted, but EMQ X is not instructed to load the code and config.